src/sources/stdin.rs

```diff
@@ -94,38 +116,52 @@ where
         let mut out =
             out.sink_map_err(|error| error!(message = "Unable to send event to out.", %error));
 
-        let res = receiver
-            .take_until(shutdown)
-            .map_err(|error| emit!(StdinReadFailed { error }))
-            .map_ok(move |line| {
-                emit!(StdinEventReceived {
-                    byte_size: line.len()
-                });
-                create_event(Bytes::from(line), &host_key, &hostname)
-            })
-            .forward(&mut out)
-            .inspect(|_| info!("Finished sending."))
-            .await;
-
-        let _ = out.flush().await; // error emitted by sink_map_err
-
-        res
-    }))
-}
+        let stream = StreamReader::new(receiver);
+        let mut stream = FramedRead::new(stream, decoder).take_until(shutdown);
+        let result = stream! {
+            loop {
+                match stream.next().await {
+                    Some(Ok((events, byte_size))) => {
+                        emit!(StdinEventsReceived {
+                            byte_size,
+                            count: events.len()
+                        });
+
+                        for mut event in events {
+                            let log = event.as_mut_log();
+
+                            log.insert(log_schema().source_type_key(), Bytes::from("stdin"));
+
+                            if let Some(hostname) = &hostname {
+                                log.insert(&host_key, hostname.clone());
+                            }
+
+                            yield event;
+                        }
+                    }
+                    Some(Err(error)) => {
+                        // Error is logged by `crate::codecs::Decoder`, no
+                        // further handling is needed here.
+                        if error.can_continue() {
+                            continue;
+                        } else {
+                            break;
+                        }
```

Comment:
Is the `continue` needed here (and in the other `Decoder` PR)? Given the position in the loop, this looks like:
```suggestion
                        if !error.can_continue() {
                            break;
                        }
```

Reply:
Changed in c5ae936d747e2e422ca369dbd929c31a2bf4013c